modules/token-swap/src/lib.rs

```diff
@@ -0,0 +1,958 @@
+		type BridgedAccountPublic: Parameter;
+		/// Account signature type at the Bridged chain.
+		type BridgedAccountSignature: Parameter;
+		/// Converter from raw hash (derived from Bridged chain account) to This chain account.
+		type FromBridgedToThisAccountIdConverter: Convert<H256, Self::AccountId>;
+	}
+
+	/// SCALE-encoded `Currency::transfer` call on the bridged chain.
+	pub type RawBridgedTransferCall = Vec<u8>;
+	/// Bridge message payload used by the pallet.
+	pub type MessagePayloadOf<T, I> = bp_message_dispatch::MessagePayload<
+		<T as frame_system::Config>::AccountId,
+		<T as Config<I>>::BridgedAccountPublic,
+		<T as Config<I>>::BridgedAccountSignature,
+		RawBridgedTransferCall,
+	>;
+	/// Type of `TokenSwap` used by the pallet.
+	pub type TokenSwapOf<T, I> = TokenSwap<
+		BlockNumberFor<T>,
+		<<T as Config<I>>::ThisCurrency as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		<T as frame_system::Config>::AccountId,
+		<T as Config<I>>::BridgedBalance,
+		<T as Config<I>>::BridgedAccountId,
+	>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Start token swap procedure.
+		///
+		/// The dispatch origin for this call must be exactly the `swap.source_account_at_this_chain` account.
+		///
+		/// Method arguments are:
+		///
+		/// - `swap` - token swap intention;
+		/// - `target_public_at_bridged_chain` - the public key of the `swap.target_account_at_bridged_chain`
+		///   account used to verify `bridged_currency_transfer_signature`;
+		/// - `bridged_currency_transfer` - the SCALE-encoded tokens transfer call at the Bridged chain;
+		/// - `bridged_currency_transfer_signature` - the signature of the `swap.target_account_at_bridged_chain`
+		///   for the message returned by the `pallet_bridge_dispatch::account_ownership_digest()` function call.
+		///
+		/// The `source_account_at_this_chain` MUST have enough balance to cover both token swap and message
+		/// transfer. Message fee may be estimated using corresponding `OutboundLaneApi` of This runtime.
+		///
+		/// **WARNING**: the submitter of this transaction is responsible for verifying:
+		///
+		/// 1) that the `bridged_currency_transfer` represents a valid token transfer call that transfers
+		///    `swap.target_balance_at_bridged_chain` to his `source_account_at_bridged_chain` account;
+		/// 2) that either the `source_account_at_bridged_chain` already exists, or the
+		///    `swap.target_balance_at_bridged_chain` is above existential deposit of the Bridged chain;
+		/// 3) the `target_public_at_bridged_chain` matches the `swap.target_account_at_bridged_chain`;
+		/// 4) the `bridged_currency_transfer_signature` is valid and generated by the owner of the
+		///    `target_public_at_bridged_chain` account.
```

Comment:
```suggestion
		///    `target_public_at_bridged_chain` account (read more about [`CallOrigin::TargetAccount`]).
```
Maybe we can refer to `CallOrigin::TargetAccount` docs for more details about the signature?